at-angular.ts

```diff
@@ -107,8 +151,36 @@ module at {
             if (target.$inject && target.$inject.length > 0) {
                 factory.$inject = target.$inject.slice(0);
             }
-            angular.module(moduleName).factory(className, factory);
+            angular.module(moduleName).factory(className || getFuncName(target), factory);
+        };
+    }
+
+    export function decorator(moduleName: string, targetProvider: string): at.IClassAnnotationDecorator {
+
+        return (targetClass: any): void => {
+
+            angular
+                .module(moduleName)
+                .config([
+                    '$provide',
+                    function($provide: angular.auto.IProvideService): void {
+
+                    $provide.decorator(targetProvider, [
+                       '$delegate',
+                        ($delegate: any): void => {
+
+                            $delegate[targetProvider] = $delegate;
+
+                            angular.extend($delegate, new targetClass());
```

Comment:
Probably `someObjectWithMethod` is a POJO - was not created from a class/constructor. I am sure that doing it this way will be ok:

```typescript
$delegate = angular.extend(
    $delegate, 
    targetClass.prototype, 
    new targetClass($delegate)
);
```